socs/agents/hwp_pmx/agent.py

```diff
@@ -260,6 +295,83 @@ def _stop_acq(self, session, params=None):
         else:
             return False, 'acq is not currently running'
 
+    def initiate_shutdown(self, session, params):
+        """
+        Initiate the shutdown of the agent.
+        """
+        self.log.warn("INITIATING SHUTDOWN")
+
+        with self.lock.acquire_timeout(10, job='shutdown') as acquired:
+            if not acquired:
+                self.log.error("Could not acquire lock for shutdown.")
+                return False, "Could not acquire lock."
+
+            self.shutdown_mode = True
+            self.dev.ign_external_voltage()
+            self.dev.turn_off()
+
+    def reverse_shutdown(self, session, params):
+        """
+        Reverses shutdown mode, allowing other tasks to update the power supply
+        """
+        self.shutdown_mode = False
+        return True, "Reversed shutdown mode"
+
+    def monitor_shutdown(self, session, params):
+        """
+        Monitor the shutdown of the agent.
+        """
+
+        session.set_status('running')
+        last_ok_time = time.time()
+
+        if self.supervisor_id is None:
+            return False, "No supervisor ID set"
+
+        while session.status in ['starting', 'running']:
+
+            res = get_op_data(self.supervisor_id, 'monitor')
+            if res['status'] != 'ok':
+                action = 'no_data'
+            else:
+                action = res['data']['actions']['rotation']
+
+            # If action is 'ok', update last_ok_time
+            if action == 'ok':
+                last_ok_time = time.time()
+
+            # If action is 'no_data', check if last_ok_time, and potentially
+            # trigger shutdown
+            elif action == 'no_data':
+                if (time.time() - last_ok_time) > self.no_data_timeout:
+                    if not self.shutdown_mode:
+                        self.agent.start('initiate_shutdown')
+
+            # If action is 'shutdown', trigger shutdown
+            elif action == 'stop':
+                if not self.shutdown_mode:
+                    self.agent.start('initiate_shutdown')
```

Comment:
You should also pass `params` here, even if just `None`. (If `initiate_shutdown` had the params decorator, this as is would run into an error.) Same goes for other direct calls to `.start()`.